### PR TITLE
Z_STEPPER_AUTO_ALIGN — To align with Z stepper points

### PR DESCRIPTION
**Pull request by sjasonsmith**
### Description

Add a new mechanism for calculating bed alignment for triple-Z printers.
Requires Z_STEPPER_AUTO_ALIGN and Z_TRIPLE_STEPPER_DRIVERS

Enabling Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS causes G34 to change in the following ways:

1. More than 3 points may be probed
2. Use a least-squares fit to calculate a plane, representing the current bed position
3. Use this plane to calculate the current height of each Z stepper
4. Proceed with the existing adjustment code, as if the bed were probed exactly over each stepper

M422 was also modified to support this functionality:

1. S parameter continues to modify one probe point
2. S now checks that points are probeable, rather than simply inside the bed dimensions
3. W parameter is used to modify one stepper location (This was arbitrarily chosen)
4. M422 with no parameters prints currently active values

#### More rapid convergence
This is most important for printers with non-ideal geometries. This was tested with a LeapFrog Xeed, which only prints on one small portion of a large bed. With the already existing G34 implementions, 10-20 G34 iterations could be required to correct a heavily skewed bed.

Selecting probe points on this printer was difficult. Probing points an equal distance to each screw required probing on a tiny portion of the bed, with very large amplification needed. Probing points near the edges of the probable area caused each screw to have very uneven contributions when adjusted.

Using Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS, I correct approximately 95% of my error in each iteration. 3-4 iterations is sufficient to align the bed. With improved tuning of Z stepper positions, I expect to align the bed in a single iteration.
 
#### Allows for additional probing of uneven surfaces
Three points is sufficient to align a perfectly flat bed. Unfortunately, many of our beds are not perfect. Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS allows additional points to be probed, which will be be least-squares fit to a plane approximating the bed. Combining this aligned plane with ABL should provide the best printing experience for uneven surfaces.

### Possible Future Improvements
- Support for dual Z steppers
- Calibration routine to fine-tune stepper positions

### Related Issues

This was mentioned in discussion in issue #13112, but it doesn't address that issue directly.
It is my opinion that issue should be closed, and is actually due to configuration errors.

### Additional Information
The following diagram visualizes the bed I am working with, on a LeapFrog Xeed printer.
The probeable area is the same as the printable area.
image

### CODE REVIEW

File: Marlin/src/gcode/calibrate/G34_M422.cpp [hunk 1]
```diff
@@ -152,17 +163,19 @@ void GcodeSuite::G34() {
       SERIAL_ECHOLNPAIR("\nITERATION: ", int(iteration + 1));
 
       // Initialize minimum value
-      float z_measured_min = 100000.0f;
+      float z_measured_min = 100000.0f,
+            z_measured_max = 0.0;
```
Comment (sjasonsmith) [inline Marlin/src/gcode/calibrate/G34_M422.cpp]: I should have initialized this to a large negative number.
```suggestion
            z_measured_max = -100000.0f;
```

**Comment by sjasonsmith** [comment]
I assume there will be some discussion and rework associated with this change.

I do not really like using 'W' as the parameter for stepper location. I think I chose this because 'S' was already used and 'W' is in screW.

I have tested the new functionality with my LeapFrog Xeed. I need to re-test to be sure the old mechanism still works, for both X_DUAL_STEPPER_DRIVERS, and X_TRIPLE_STEPPER_DRIVERS.

I do not have a printer with dual steppers, but I should be able to fake it enough to test that I did not break Z_STEPPER_AUTO_ALIGN.

**Comment by sjasonsmith** [comment]
I re-tested this code today on my Leapfrog Xeed, to verify that my changes do not prevent Z_STEPPER_AUTO_ALIGN from working without Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS.

For both tests, I start with a level bed. I then dropped the rear screw 5mm, and the left screw 3mm. I then performed a single 3-iteration Z_STEPPER_AUTO_ALIGN to compare the maximum difference between measurements prior to each round of adjustments.

It is important to note that with Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS, the differences shown are the calculated differences at the screw X/Y positions, rather than the actual probe heights.

| Measurement Iteration | Without Known Stepper Positions (Max Diff at Probe X/Y) | With Known Stepper Positions (Max Diff at Screw X/Y) |
|-----------------------|---------------------------------|------------------------------|
| 1 | 1.79 mm | 4.98 mm |
| 2 | 0.96 mm | 0.20 mm |
| 3 | 0.35 mm | 0.04 mm |

**Comment by sjasonsmith** [comment]
I did not have time to verify existing behavior still works with Dual Z motors. If anyone is interested in trying it out on a dual Z printer to ensure I didn't break anything, I would much appreciate it.

The mechanism could be modified to work with Dual Z, but I'm not sure how important it is. I suspect the existing amplification factor could be tuned to get good behavior on a Dual Z printer.

**Comment by sjasonsmith** [comment]
I configured my printer to only use two of the screws to mimic a Dual Z printer, and verified that the old alignment mechanism still works with it.

**Comment by sjasonsmith** [comment]
I'd be happy to retarget this to dev-2.1.x. I hadn't originally because it seemed far behind the current state of bugfix-2.0.x, and I wasn't sure whether it was being actively used yet.

**Comment by thinkyhead** [comment]
We can leave this where it is, and after the 2.0 release we can move it if needed, or do a "side merge" where we merge it into the other branch directly and just close this at that time.

**Comment by sjasonsmith** [comment]
I’m not expecting further input on this until after 2.0 is done. I’m pushing changes to it as I reconcile with other recent changes in bf2, but there is still work to be done in that regard.

File: Marlin/Configuration_adv.h [hunk 2]
```diff
@@ -613,7 +637,9 @@
   // Use the amplification factor to de-/increase correction step.
   // In case the stepper (spindle) position is further out than the test point
   // Use a value > 1. NOTE: This may cause instability
+  // Must be 1.0 when using Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS
```
Comment (InsanityAutomation) [inline Marlin/Configuration_adv.h]: Put this inside an #if DISABLED( block then in conditionals post use #if ENABLED on Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS and force it to 1.0

File: Marlin/src/gcode/calibrate/G34_M422.cpp [hunk 3]
```diff
@@ -147,11 +171,11 @@ void GcodeSuite::G34() {
...
 
     float last_z_align_move[Z_STEPPER_COUNT] = ARRAY_N(Z_STEPPER_COUNT, 10000.0f, 10000.0f, 10000.0f),
-          z_measured[Z_STEPPER_COUNT] = { 0 },
-          z_maxdiff = 0.0f,
-          amplification = z_auto_align_amplification;
-
-    const ProbePtRaise raise_after = parser.boolval('E') ? PROBE_PT_STOW : PROBE_PT_RAISE;
+          z_measured[G34_PROBE_COUNT] = { 0 },
+          z_maxdiff = 0.0f;
+    //#if DISABLED(Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS)
```
Comment (InsanityAutomation) [inline Marlin/src/gcode/calibrate/G34_M422.cpp]: With the above, the commented code can go away

File: Marlin/src/gcode/calibrate/G34_M422.cpp [hunk 4]
```diff
@@ -180,35 +206,59 @@ void GcodeSuite::G34() {
...
 
-      SERIAL_ECHOPAIR("\n"
+      SERIAL_ECHOLNPAIR("\n"
         "DIFFERENCE Z1-Z2=", ABS(z_measured[0] - z_measured[1])
         #if ENABLED(Z_TRIPLE_STEPPER_DRIVERS)
           , " Z2-Z3=", ABS(z_measured[1] - z_measured[2])
           , " Z3-Z1=", ABS(z_measured[2] - z_measured[0])
         #endif
       );
       SERIAL_EOL();
```
Comment (InsanityAutomation) [inline Marlin/src/gcode/calibrate/G34_M422.cpp]: Both SERIAL_EOL();'s can go with that being an EchoLN above

**Comment by InsanityAutomation** [comment]
Aside from the review comments added, example configs will need to be updated. I would clarify the verbiage there a bit more though. I can see many users being very confused what to put.

**Comment by InsanityAutomation** [comment]
@thinkyhead Would you mind considering this for a sooner application? The more I look at the quad Z we discussed at ERRF the more it overlaps with this. Also as for variable naming, for consistency with with RRF, should we call it Ballscrew positions instead? I know it could be leadscrew or belt instead. Stepper may not be accurate either. Possibly Z drive or rail as a generic? Consistent to RRF is probably best for now though...

**Comment by sjasonsmith** [comment]
> Aside from the review comments added, example configs will need to be updated. I would clarify the verbiage there a bit more though. I can see many users being very confused what to put.

I expect to have to revisit this once 2.0 stabilizes and is released. I would change it back to draft mode, but I don't see a way to do that. I didn't want to spend time rolling it into every example without first receiving feedback that the approach would be acceptable. I'd be thrilled to receive feedback and iterate on this to get it where you want it.

> The more I look at the quad Z we discussed at ERRF the more it overlaps with this.

Are you thinking of a shared implementation for both, or you just want to standardize naming on something that will work for both? I believe Klipper has separate algorithms for three and four motors, I don't know about RRF.

The approach I'm using doesn't extend to four motors, since I calculate the height of each Z motor based on an ideal flat plane. I assume that plane remains flat as it is adjusted, which isn't true with four motors. That allows a finely-tuned 3-screw system to converge in a single iteration with some fairly simple math.

The actual point definition and probing portions of it could of course be shared, even if they use a different algorithm to calculate the needed movement.

> Also as for variable naming, for consistency with with RRF, should we call it Ballscrew positions instead?

I originally called everything screws, and made it more generic to avoid offending people with belts.
My printer has ball screws, so it doesn't bother me. :)

**Comment by InsanityAutomation** [comment]
Primarily naming conventions and we would be tripping over the same areas of the same files most likely resulting in a bunch of merge conflicts, especially with the configuration files. The actual implementations in both klipper and reprap were written by the voron design guys and I've been in some discussion with them. They have a legacy implementation they had written for Marlin and we're trying to track down the original source for me to work on porting to the current branches. If they can't locate the Old source though I will end up going over klipper and essentially porting that methodology since it should be what they are referencing from now days anyway.

**Comment by sjasonsmith** [comment]
If you have recommendations for name changes I can go ahead and make those changes, along with the other suggestions you've made. I'll be able to make changes and actually test them this week, but will be unavailable next week.

File: Marlin/src/gcode/calibrate/G34_M422.cpp [hunk 5]
```diff
@@ -110,11 +110,15 @@ void GcodeSuite::G34() {
...
-    if (!WITHIN(ABS(z_auto_align_amplification), 0.5f, 2.0f)) {
-      SERIAL_ECHOLNPGM("?(A)mplification out of bounds (0.5-2.0).");
-      break;
-    }
+    const float z_auto_align_amplification =
+      #if ENABLED(Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS)
+        Z_STEPPER_ALIGN_AMP;
+      #else
+        parser.floatval('A', Z_STEPPER_ALIGN_AMP);
+        if (!WITHIN(ABS(z_auto_align_amplification), 0.5f, 2.0f)) {
```
Comment (InsanityAutomation) [inline Marlin/src/gcode/calibrate/G34_M422.cpp:112]: Close } appears to be missing here

Comment (sjasonsmith) [reply Marlin/src/gcode/calibrate/G34_M422.cpp:112]: Oops, missed a commit when I pushed those changes.

**Comment by thinkyhead** [comment]
I don't mind bringing this in (if I can make it pretty enough). I'm planning a big overhaul for 2.1 that involves changing the way we deal with assigning axes to stepper motors / drivers, so it seems good to get this in in the "old style" and include it in 2.0 with the other multi-stepper-per-axis options.

So, let me give this a once-over, and it can get merged shortly thereafter.

**Comment by thinkyhead** [comment]
Rebased and squashed. To get the latest into your working copy:
```
git fetch origin
git checkout G34-KnownStepperPositions
git reset --hard origin/G34-KnownStepperPositions
```

**Comment by sjasonsmith** [comment]
Thanks @thinkyhead. Let me know anything you want me to change or test with it. 

I have changes in progress to update config comments and to let it auto-select three points to probe if none are defined, to simplify a first-time configuration and be more in-line with @InsanityAutomation's recent leveling changes, including adapting to probe offset changes.

I don't want to overlap with any changes you are making now, so I'll hold off on pushing anything further until you say so, or I can post as a separate pull request after this goes in, if you prefer.

**Comment by thinkyhead** [comment]
I just had to patch for the updated position types, but otherwise I left the logic as it was conceived. Give the code a proper test, fix anything that seems broken. If you have a machine that matches one of our example configurations, you can also apply your settings to the example and submit that for others to use.

**Comment by sjasonsmith** [comment]
> If you have a machine that matches one of our example configurations, you can also apply your settings to the example and submit that for others to use.

My machine doesn't even have a compatible board in the mainline right now, adding that will probably show up in a PR eventually. It's Leapfrog's "big" board, so I doubt there are a lot of users clamoring for it.

**Comment by phongshader** [comment]
I have a question on where the Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS should be measured from/to. I have a bed that is 535x535 with a printable area of 490x465. Z1 is -95,0 Z2 is 565,0 (+75 on x) and Z3 is 230, 645(+180 on y).  I think what you're saying is that the probe points should within printable area and all on the same radius from the center of the circle created by the z axes...right?

**Comment by sjasonsmith** [comment]
@phongshader if you are using know stepper positions just let it pick points for you. It doesn’t matter where they are or in what order, as long as your bed is flat. They just have to form a triangle and not a straight line.

This algorithm just uses them to calculate a flat plane and extrapolate that plane out to your specified screw positions. Ideally the points would be as far apart as possible, and avoid any bad areas in your bed if you have divots or bumps on the surface.

**Comment by phongshader** [comment]
@sjasonsmith  I've tried to get it working but every time I run a G34 command the level gets worse so I'm trying to figure out why. The Z_STEPPER_ALIGN_KNOWN_STEPPER_POSITIONS in my configuration_adv file are WAY OFF but I put them in before reading this. But even before putting those positions in the leveling never worked for me. When you say " just let it pick points for you" does that mean I should comment out #define Z_STEPPER_ALIGN_STEPPER_XY? That fails when trying to compile. Or the better question is how do I let it pick for me?
[configs.zip](https://github.com/MarlinFirmware/Marlin/files/7306693/configs.zip)

**Pull request merged**